src/util/Scheduler.h

```diff
@@ -0,0 +1,204 @@
+//     to be suddenly full of ready actions, or continuously-reschedule itself,
+//     so we make sure no queue can have less than some (steadily rising) floor.
+//
+//   - We encode deadlines in actions: those with a positive deadline are always
+//     dropped -- unconditionally -- if they are ready to run after their
+//     deadline, and may also be dropped conditionally if the system is under
+//     load / the queue is too long. The sentinel "minimum deadline" value is
+//     reserved to indicate a never-droppable action. To encode a "best effort"
+//     action with no particular deadline that will be dropped only under load,
+//     we set the deadline to the the maximal duration value.
+
+namespace stellar
+{
+
+using Action = std::function<void()>;
+
+class Scheduler
+{
+  public:
+    struct Stats
+    {
+        size_t mActionsEnqueued{0};
+        size_t mActionsDequeued{0};
+        size_t mActionsDroppedDueToOverload{0};
+        size_t mActionsDroppedDueToDeadline{0};
+        size_t mQueuesActivatedFromFresh{0};
+        size_t mQueuesActivatedFromCache{0};
+        size_t mQueuesSuspended{0};
+    };
+
+    using RelativeDeadline = std::chrono::nanoseconds;
+    using AbsoluteDeadline = std::chrono::steady_clock::time_point;
+    static constexpr RelativeDeadline NEVER_DROP = RelativeDeadline::min();
+    static constexpr RelativeDeadline DROP_ONLY_UNDER_LOAD =
+        RelativeDeadline::max();
+    static constexpr AbsoluteDeadline ABS_NEVER_DROP = AbsoluteDeadline::min();
+    static constexpr AbsoluteDeadline ABS_DROP_ONLY_UNDER_LOAD =
+        AbsoluteDeadline::max();
+
+  private:
+    class Queue;
+    using Qptr = std::shared_ptr<Queue>;
+    std::map<std::string, Qptr> mQueues;
+    std::priority_queue<Qptr, std::vector<Qptr>,
+                        std::function<bool(Qptr, Qptr)>>
+        mQueueQueue;
+
+    Stats mStats;
+
+    // A queue is considered "overloaded" if its size is above the load limit.
+    // This is a per-queue limit.
+    size_t const mLoadLimit;
+
+    // The serviceTime of any queue will always be advanced to at least this
+    // duration behind mMaxServiceTime, to limit the amount of "suplus" service
+    // time any given queue can acucmulate if it happens to go idle a long time.
+    std::chrono::nanoseconds const mServiceTimeWindow;
+
+    // Largest serviceTime seen in any queue. This number will continuously
+    // advance as queues are serviced; it exists to serve as the upper limit
+    // of the window, from which mServiceTimeWindow is subtracted to derive
+    // the lower limit.
+    std::chrono::nanoseconds mMaxServiceTime{0};
+
+    // Sum of sizes of all the active queues. Maintained as items are enqueued
+    // or run.
+    size_t mSize{0};
+
+    // We cache recent queues for a while after we empty them, so that we can
+    // maintain a service-level estimate spanning their repeated activations.
+    RandomEvictionCache<std::string, Qptr> mQueueCache{1024};
```

Comment:
Well, if they're idle we have to put them _somewhere_ that's not the runnable queue, otherwise they'll quickly move to the front of the runnable queue -- since they're not accumulating new service time -- and sit there not-accumulating service time.

The cache was just a way to keep them around over idle periods so we mostly preserve accounting information. I can try to rework this to be time based.

Reply:
(IOW in the time between made-idle and idle-so-long-we-need-to-drop, they need to be stored somewhere and that somewhere has to allow us to make them non-idle in O(small) and allow us to drop the longest-idle one in O(small). I'd use a priority queue for _that_ also but you can't pull random elements out of the middle of a priority queue, alas. I think probably the only way to do this is to make a secondary intrusive doubly-linked list running through the ActionQueues indicating their position in the idleness-order. Which is a bit gross but maybe not more gross than the random eviction cache, idk. I'll do a sketch to show what it looks like.)

Reply:
Fixed.